Explore simulated data

ajlee21 opened the pull request:
This PR adds 3 scripts to explore the simulated data in order to try to explain why we do not see a difference in the prediction results using nonlinear compression (VAE) compared to linear compression (PCA) and no compression (gene space).

1.  explore_simulated_data:  Look at the simulated data created.  Do we see the signal that we've artifically added?
2.  explore_HighWeight_genes:  Look at the VAE latent space features, are they capturing the signal from A and B?
3.   explore_offset:  Look at the offset vectors used to perform the transformation of the data, is it capturing the signal from A and B?

Other files re-generated from re-running the scripts during the exploration.

gwaybio commented on data/sim_AB_2775_300_v2/geneSetA.txt:
```diff
@@ -0,0 +1,101 @@
+	0
```
these types of files should always have column names

ajlee21 replied:
added column name to both A and B gene list files

gwaybio commented on scripts/functions/interpolate.py:
```diff
@@ -213,18 +213,23 @@ def interpolate_in_vae_latent_space(data_dir, model_dir, encoded_dir, latent_dim
         alpha = utils.get_scale_factor(
             target_gene_sorted, gene_id, intermediate_target_gene_exp, percent_low, percent_high)
 
+        #print("baseline is {}".format(baseline_encoded))
```
remove debugging print statements!

ajlee21 replied:
removed

gwaybio commented on scripts/functions/utils.py:
```diff
@@ -46,6 +46,8 @@ def get_scale_factor(target_gene_sorted, gene_id, expression_profile,
     # scale_factor is the proportion along the gene expression gradient
     scale_factor = expression_profile / denom
 
+    #print("Denominator is {}".format(denom))
```
remove print statements here too

ajlee21 replied:
removed

gwaybio commented on scripts/functions/vae.py:
```diff
@@ -272,3 +272,26 @@ def tybalt_2layer_model(learning_rate, batch_size, epochs, kappa, intermediate_d
+    hidden_weight_df = pd.DataFrame(weights[1][2])
+    abstracted_weight_df = intermediate_weight_df.dot(hidden_weight_df)
+
+    abstracted_weight_df.index = range(1, latent_dim + 1)
```
unless you've been consistent with 1 indexing for this project, I would avoid doing this here.

I did it in the original Tybalt repo, but I should have stuck with zero indexing!

ajlee21 replied:
This is a good catch.  The other column labels for the latent space dimension (that I use in other scripts) starts with index 0.

gwaybio commented on scripts/functions/vae.py:
```diff
@@ -272,3 +272,26 @@ def tybalt_2layer_model(learning_rate, batch_size, epochs, kappa, intermediate_d
+    # Multiply hidden layers together to obtain a single representation of gene weights
+    intermediate_weight_df = pd.DataFrame(weights[1][0])
+    hidden_weight_df = pd.DataFrame(weights[1][2])
+    abstracted_weight_df = intermediate_weight_df.dot(hidden_weight_df)
```
I think it is worth considering the limitations of this multiplication. Simply multiplying the weight matrices together ignores the activation function between layers. It may be worth mentioning, and perhaps even outputing both matrices in this function

ajlee21 replied:
Very good point.  This point was brought up by our collaborator, who was curious about the contribution of genes in the first layer versus the second layer.

ajlee21 replied:
This is added as issue https://github.com/greenelab/Pseudomonas_latent_spaces/issues/10

gwaybio commented on scripts/nbconverted/explore_HighWeight_genes.py:
```diff
@@ -0,0 +1,242 @@
+
+
+# Plot the distribution of gene weights per latent feature
+#num_features = weight.shape[1]
```
why commented?

ajlee21 replied:
Here I was trying to visualize the distribution of weights for all 300 latent space features, but I had trouble trying to figure out the plot something so large.  I think its something I would like to still look into in the future, which is why it is commented here instead of just being deleted.  It was taking a bit too long and I felt that my time might be better spent on some of the other analyses temporarily

gwaybio commented on scripts/nbconverted/explore_HighWeight_genes.py:
```diff
@@ -0,0 +1,242 @@
+    pos_threshold = node_mean + two_stds.iloc[i]
+    neg_threshold = node_mean - two_stds.iloc[i]
+    
+    hw_pos_genes = weight[weight[i+1] > pos_threshold].index
```
if you decide to zero index weights, this will need to be updated too

ajlee21 replied:
Very good eye, I have updated this

gwaybio commented on scripts/nbconverted/explore_input_rnaseq.py:
```diff
@@ -0,0 +1,85 @@
+# In[4]:
+
+
+## 1. Distribution of PA1673 across samples
```
Why are we interested in PA1673?

ajlee21 replied:
Tried to add comments to explain this in the notebook

gwaybio commented on scripts/nbconverted/explore_offset.py:
```diff
@@ -0,0 +1,259 @@
+max_value = max(offset_vae_space.T[0])
+min_value = min(offset_vae_space.T[0])
+
+max_feature = offset_vae_space.T[offset_vae_space.T[0] == max_value].index[0]
```
checkout [pd.DataFrame().idxmin](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.idxmin.html)

There is also `idxmax`

ajlee21 replied:
good tip

gwaybio commented on scripts/nbconverted/explore_offset.py:
```diff
@@ -0,0 +1,259 @@
+
+
+# Get gene ids with the highest negative weight from the min feature selected
+percentile = 5
```
The Venn plots and the percentile gene discovery could probably be made into a function

gwaybio replied:
Will reduce amount of code

ajlee21 replied:
Will address in the next PR

ajlee21 replied:
Issue https://github.com/greenelab/Pseudomonas_latent_spaces/issues/11

gwaybio commented on scripts/nbconverted/explore_HighWeight_genes.py:
```diff
@@ -0,0 +1,242 @@
+    add_metadata_neg = pd.DataFrame({node_name: [node_mean, node_std, neg_threshold]})
+    HW_metadata_df = pd.concat([HW_metadata_df, add_metadata_neg], axis=1)
+
+HW_df = HW_df.T
```
Why are there a lot of NAs in this dataframe?

What is the structure of this data? What will this data be used for downstream? It is important to consider these things when creating the dataframe. We can circle back to the best way to present this data after you determine how you will use it

ajlee21 replied:
This is a dataframe that was created to be uploaded into the Greenelab ADAGE server (https://adage.greenelab.com/#/home).  This is the format that Dongbo has setup for the database.

gwaybio commented on scripts/nbconverted/explore_HighWeight_genes.py:
```diff
@@ -0,0 +1,242 @@
+    HW_metadata_df = pd.concat([HW_metadata_df, add_metadata_neg], axis=1)
+
+HW_df = HW_df.T
+HW_metadata_df = HW_metadata_df.T
```
Same thing goes with this dataframe. I also think we can avoid the for loop here and just use pandas magic. Something like:

```python
weight = weight.assign(hw_mean=weight.mean(axis=0),
                       hw_std=weight.std(axis=0))

threshold_high_df = weight.hw_mean + (2 * weight.std)
threshold_low_df = weight.hw_mean - (2 * weight.std)

HW_metadata_df = pd.concat([weight.loc[:, ['hw_mean', 'hw_std']],
                            threshold_high_df,
                            threshold_low_df])
```

ajlee21 replied:
Very helpful tip -- thank you for the advice, I definitely need to learn more of the fancy data structures available in pandas.  Will address in the next PR

gwaybio replied:
I think you've made an issue that corresponds with this comment? it is good to link to the issue in the comment (#10 i believe?)

ajlee21 replied:
Thank you for explaining how to link issues.  This is for issue https://github.com/greenelab/Pseudomonas_latent_spaces/issues/11

gwaybio commented on scripts/nbconverted/explore_HighWeight_genes.py:
```diff
@@ -0,0 +1,242 @@
+
+Weights_pos_neg_df = pd.DataFrame()
+
+for i in range(num_nodes):
```
Might be possible to avoid the loop here. can you just subtract the mean from the dataframe directly? See [pandas.DataFrame.subtract](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.subtract.html)

ajlee21 replied:
Will address in the next PR

gwaybio replied:
Is this #11?

ajlee21 replied:
This is also issue https://github.com/greenelab/Pseudomonas_latent_spaces/issues/11

gwaybio commented on scripts/nbconverted/explore_HighWeight_genes.py:
```diff
@@ -0,0 +1,242 @@
+                       })
+    percent_overlap = percent_overlap.append(add)
+
+percent_overlap
```
Why do we care about the percent overlap? If the geneset is in the feature then we can say something about that geneset or that feature? or is this to nominate/prioritize features?

Also, do you want to sort this to highlight the nodes that have the highest percent overlap?

ajlee21 replied:
This was just a very first pass at exploring the data.  Based on the results of the interpolation analysis we were NOT seeing a discrepancy between the ability for VAE vs PCA vs No compression models to predict intermediate gene expression levels as gene set A is becoming activated and gene set B is getting upregulated.  So we first were wondering if genes and A and B were being captured by our method.  I guess I could've made it binary to look for presence or absence, but I didn't think of that at the time.  Instead I just looked at percentages which gave me information about "how much" of the genes in A or B were captured in addition to their presence.  But there is no ranking being done here.

gwaybio commented on scripts/nbconverted/explore_input_rnaseq.py:
```diff
@@ -0,0 +1,85 @@
+# In[8]:
+
+
+get_ipython().run_cell_magic('time', '', 'from statsmodels.nonparametric.smoothers_lowess import lowess\n## 4.  Plot the expression of genes along the PA1673 gradient:\n# Are the genes changing linearly with respect to PA1673?\n\n# Sort samples by PA1673 expression (lowest --> highest)\ndata_sorted = data.sort_values(by=[gene_id])\ndata_sorted.head()\n\n# For each gene plot gene expression trend  ...
```
imports should go in the first code cell

Probably don't need to time the whole execution, and just the `for panel in range(...)` bit, right?

gwaybio replied:
you have a note that it is not incrementing correctly? Is it now?

gwaybio replied:
I see the purpose of this now! We are searching for non linear relationships? Pretty cool. We do see a bunch! There are a lot of wild patterns.

Is there a better way to search for non linear patterns than just by eye?

gwaybio replied:
Also, it looks like the axis labels on these plots are a bit wonky

ajlee21 replied:
1. Moved the import statement
2.  Yes the note was when I initially wrote the script and I realized at the end it wasn't indexing properly but it should be now
3.  When looking for nonlinear patterns, the only think I could think to do is 1) Do a correlation analysis to see which genes were Linearly correlated vs not and 2) To visualize them all because the ones that are not linearly correlated may mostly be noise but there may be some with interesting patterns and those are the ones we care about.  This is something I'll need to think a bit more about -- maybe doing some sort of smoothing and somehow quantifying the trend rule out the "noisy" patterns but need to think about the actual implementation here.

gwaybio replied:
ah, gotcha. Cool.

Could you test all relationships for a series of correlations? Maybe a Pearson and Spearman correlation plus some sort of non linear curve fitting. Maybe some combinations of values for these three scores could help prioritize which gene-gene relationships to focus on?

ajlee21 replied:
Very good idea.  Added as issue: https://github.com/greenelab/Pseudomonas_latent_spaces/issues/12

gwaybio commented on scripts/nbconverted/explore_offset.py:
```diff
@@ -0,0 +1,259 @@
+# In[31]:
+
+
+# Compare the overlap of genes in set B and highest negative weighted genes in the min feature
```
The min and max features have the same overlap? This is suspicious to me

ajlee21 replied:
I noticed this as well, but the analysis was a bit rushed.  Will need to look back into it in the next PR

gwaybio replied:
#11 ?

ajlee21 replied:
Issue https://github.com/greenelab/Pseudomonas_latent_spaces/issues/13

ajlee21 commented:
Think I have addressed all your comments.  There are a few that I have made into issues for future PRs.  This was a very rough analysis just to start exploring the data and you've pointed out very good areas of improvement moving forward.  Please let me know if you want to discuss any of the issues further, especially in terms of research strategy
